### ISSUE TITLE
Fix for infinite sync of carts with unsupported products and emptied carts not being removed

### ISSUE THREAD

**Issue by TroyPatteson** [mailchimp/mc-magento#1032]
As the title indicates there are two use cases that are not correctly handled in the current development version of mc-magento since the release of v1.1.17 and probably earlier versions:

1) Adding only unsupported products to a new quote.

Expected behaviour:

The quote is not synchronised to mailchimp but marked as having been processed so that it is no longer considered until it is next modified.

Actual behaviour:

The quote is attempted to be synchronised every time the cron runs until either the cart is emptied, a supported product is added or the quote is converted to an order where it is no longer active.

Diagnosis:

The query to get all new quotes ignores any new active quotes containing no items but cannot ignore quotes containing only unsupported products such as bundled products. Since unsupported products cannot be synced to Mailchimp the quote is ignored and subsequently returned next time the query to get all new quotes is run.

2) Emptying a cart containing supported products

Expected behaviour:

The quote is removed from mailchimp.

Actual behaviour:

The quote is not removed from mailchimp.

Proposal:

To stop 1) rather than just ignoring the cart which results in it being processed as a new cart next time, add a sync item to indicate it has been synced but with the _mailchimp_sync_deleted_ field set to 1 to indicate that there is no record on mailchimp. Since there is no associated batch operation set the batch ID to "ignored" to prevent the sync item from being treated as an unsent item and deleted when the next store is synced.

To make these changes first change the method _saveEcommerceSyncData()_ of the **app/code/community/Ebizmarts/MailChimp/Helper/Data.php** helper from:
```
    /**
     * Save entry for ecommerce_sync_data table overwriting old item if exists or creating a new one if it does not.
     *
     * @param $itemId
     * @param $itemType
     * @param $mailchimpStoreId
     * @param null $syncDelta
     * @param null $syncError
     * @param int $syncModified
     * @param null $syncDeleted
     * @param null $token
     * @param null $syncedFlag
     * @param bool $saveOnlyIfexists
     * @param null $deletedRelatedId
     * @param bool $allowBatchRemoval

     */
    public function saveEcommerceSyncData(
        $itemId,
        $itemType,
        $mailchimpStoreId,
        $syncDelta = null,
        $syncError = null,
        $syncModified = 0,
        $syncDeleted = null,
        $token = null,
        $syncedFlag = null,
        $saveOnlyIfexists = false,
        $deletedRelatedId = null,
        $allowBatchRemoval = true

    ) {
        $ecommerceSyncDataItem = $this->getEcommerceSyncDataItem($itemId, $itemType, $mailchimpStoreId);
        if (!$saveOnlyIfexists || $ecommerceSyncDataItem->getMailchimpSyncDelta()) {
            if ($syncDelta) {
                $ecommerceSyncDataItem->setData("mailchimp_sync_delta", $syncDelta);
            } elseif ($allowBatchRemoval) {
                $ecommerceSyncDataItem->setData("batch_id", null);
            }

            if ($syncError) {
                $ecommerceSyncDataItem->setData("mailchimp_sync_error", $syncError);
            }
```
to
```
    /**
     * Save entry for ecommerce_sync_data table overwriting old item if exists or creating a new one if it does not.
     *
     * @param $itemId
     * @param $itemType
     * @param $mailchimpStoreId
     * @param null $syncDelta
     * @param null $syncError
     * @param int $syncModified
     * @param null $syncDeleted
     * @param null $token
     * @param null $syncedFlag
     * @param bool $saveOnlyIfexists
     * @param null $deletedRelatedId
     * @param bool $allowBatchRemoval
     * @param bool $invalidOrEmpty
     */
    public function saveEcommerceSyncData(
        $itemId,
        $itemType,
        $mailchimpStoreId,
        $syncDelta = null,
        $syncError = null,
        $syncModified = 0,
        $syncDeleted = null,
        $token = null,
        $syncedFlag = null,
        $saveOnlyIfexists = false,
        $deletedRelatedId = null,
        $allowBatchRemoval = true,
        $invalidOrEmpty = false
    ) {
        $ecommerceSyncDataItem = $this->getEcommerceSyncDataItem($itemId, $itemType, $mailchimpStoreId);
        if (!$saveOnlyIfexists || $ecommerceSyncDataItem->getMailchimpSyncDelta()) {
            if ($invalidOrEmpty) {
                /*
                 * Item is invalid or empty so ensure batch ID is not null to
                 * prevent it from being deleted as an unsent item in future.
                 */
                if (!$syncDelta) {
                    $dateHelper = $this->getDateHelper();
                    $syncDelta = $dateHelper->formatDate(null, 'Y-m-d H:i:s');
                }
                $ecommerceSyncDataItem->setData("mailchimp_sync_delta", $syncDelta);
                if ($ecommerceSyncDataItem->getData("batch_id") === null) {
                    $ecommerceSyncDataItem->setData("batch_id", "ignored");
                }
            } elseif ($syncDelta) {
                $ecommerceSyncDataItem->setData("mailchimp_sync_delta", $syncDelta);
            } elseif ($allowBatchRemoval) {
                $ecommerceSyncDataItem->setData("batch_id", null);
            }

            if ($syncError !== null) {
                $ecommerceSyncDataItem->setData("mailchimp_sync_error", $syncError);
            }
```
This change adds the parameter _$invalidOrEmpty_ which makes sure the batch ID and _mailchimp_sync_delta_ fields are not null avoiding the default _$allowBatchRemoval_ condition of setting batch ID to NULL which then results in the item being deleted next time as an unsent item.

There is a second minor change to always set _mailchimp_sync_error_ if the _$syncError_ parameter is not NULL to ensure that any earlier error is cleared when _$syncError_ is set to '' if an item is subsequently synchronised successfully.

Next modify __updateSyncData()_ method of **app/code/community/Ebizmarts/MailChimp/Model/Api/Carts.php** from:
```
    /**
     * update product sync data
     *
     * @param $cartId
     * @param $mailchimpStoreId
     * @param int|null $syncDelta
     * @param int|null $syncError
     * @param int|null $syncModified
     * @param int|null $syncedFlag
     * @param int|null $syncDeleted
     * @param string|null $token

     */
    protected function _updateSyncData(
        $cartId,
        $mailchimpStoreId,
        $syncDelta = null,
        $syncError = null,
        $syncModified = 0,
        $syncedFlag = null,
        $syncDeleted = null,
        $token = null

    ) {
        $helper = $this->getHelper();
        $helper->saveEcommerceSyncData(
            $cartId,
            Ebizmarts_MailChimp_Model_Config::IS_QUOTE,
            $mailchimpStoreId,
            $syncDelta,
            $syncError,
            $syncModified,
            $syncDeleted,
            $token,
            $syncedFlag
        );
    }
```
to:
```
    /**
     * update product sync data
     *
     * @param $cartId
     * @param $mailchimpStoreId
     * @param int|null $syncDelta
     * @param int|null $syncError
     * @param int|null $syncModified
     * @param int|null $syncedFlag
     * @param int|null $syncDeleted
     * @param string|null $token
     * @param bool $invalidOrEmpty
     */
    protected function _updateSyncData(
        $cartId,
        $mailchimpStoreId,
        $syncDelta = null,
        $syncError = null,
        $syncModified = 0,
        $syncedFlag = null,
        $syncDeleted = null,
        $token = null,
        $invalidOrEmpty = false
    ) {
        $helper = $this->getHelper();
        $helper->saveEcommerceSyncData(
            $cartId,
            Ebizmarts_MailChimp_Model_Config::IS_QUOTE,
            $mailchimpStoreId,
            $syncDelta,
            $syncError,
            $syncModified,
            $syncDeleted,
            $token,
            $syncedFlag,
            false,
            null,
            !$invalidOrEmpty,
            $invalidOrEmpty
        );
    }
```

Finally modify __getNewQuotes()_ method of **app/code/community/Ebizmarts/MailChimp/Model/Api/Carts.php** from:
```
            $cartJson = $this->_makeCart($cart, $mailchimpStoreId, $magentoStoreId);
            if ($cartJson != "") {
                $helper->modifyCounterSentPerBatch(Ebizmarts_MailChimp_Helper_Data::QUO_NEW);

                $counter = $this->getCounter();
                $allCarts[$counter]['method'] = 'POST';
                $allCarts[$counter]['path'] = '/ecommerce/stores/' . $mailchimpStoreId . '/carts';
                $allCarts[$counter]['operation_id'] = $batchId . '_' . $cartId;
                $allCarts[$counter]['body'] = $cartJson;
                $this->setCounter($this->getCounter()+1);
                $this->_updateSyncData(
                    $cartId,
                    $mailchimpStoreId,
                    null,
                    null,
                    0,
                    null,
                    null,
                    $this->getToken()
                );
                /*
                $helper->logDebug(
                    "Creating e-commerce batch for store $magentoStoreId: "
                    . "Adding new cart ID $cartId for customer email "
                    . $cart->getCustomerEmail() . " currency "
                    . $cart->getQuoteCurrencyCode() . " order_total "
                    . $cart->getGrandTotal()
                );
                */
            } else {
                $this->_updateSyncData($cartId, $mailchimpStoreId);
            }

            $this->setToken(null);
        }

        return $allCarts;
```
to:
```
            $cartJson = $this->_makeCart($cart, $mailchimpStoreId, $magentoStoreId);
            if ($cartJson != "") {
                $helper->modifyCounterSentPerBatch(Ebizmarts_MailChimp_Helper_Data::QUO_NEW);

                $counter = $this->getCounter();
                $allCarts[$counter]['method'] = 'POST';
                $allCarts[$counter]['path'] = '/ecommerce/stores/' . $mailchimpStoreId . '/carts';
                $allCarts[$counter]['operation_id'] = $batchId . '_' . $cartId;
                $allCarts[$counter]['body'] = $cartJson;
                $this->setCounter($this->getCounter()+1);
                $this->_updateSyncData(
                    $cartId,
                    $mailchimpStoreId,
                    null,
                    null,
                    0,
                    null,
                    null,
                    $this->getToken()
                );
                /*
                $helper->logDebug(
                    "Creating e-commerce batch for store $magentoStoreId: "
                    . "Adding new cart ID $cartId for customer email "
                    . $cart->getCustomerEmail() . " currency "
                    . $cart->getQuoteCurrencyCode() . " order_total "
                    . $cart->getGrandTotal()
                );
                */
            } else {
                /*
                 * Item contains only unsupported products. Mark it as
                 * unsupported so that the item is not deleted with any other
                 * unsent items (batch ID null) or it will synchronise
                 * indefinitely.
                 */
                /*
                $helper->logWarning(
                    "Creating e-commerce batch for store $magentoStoreId: "
                    . "Ignoring cart ID {$cart->getId()} for customer email "
                    . $cart->getCustomerEmail() . ": no supported products"
                );
                */
                $this->_updateSyncData(
                    $cartId,
                    $mailchimpStoreId,
                    null,
                    null,
                    0,
                    null,
                    1,
                    null,
                    true
                );
            }

            $this->setToken(null);
        }

        return $allCarts;
```

To solve 2) the query to get modified quotes must be changed to also get active quotes whose sync item is marked as deleted as any subsequent cart modification could be to add supported products to the cart. Then when processing the modified cart the following conditions must be handled:

a) If the cart was not previously marked as deleted indicating that it exists in mailchimp and the cart was modified such that it is now empty or only contains unsupported products then the cart must be deleted and the quote sync item marked as deleted.

b) If the cart was previously marked as deleted indicating that it does not exist in mailchimp and the cart was modified such that it is now empty or only contains unsupported products then the quote sync item is marked as invalid or empty (batch ID set to "ignored") so that it is not deleted as an unsent item.

c) If the cart was not previously marked as deleted indicating that it exists in mailchimp and the cart was modified such that it contains one or more supported products then the cart in mailchimp must be updated with a PATCH request.

d) If the cart was previously marked as deleted indicating that does not exist in mailchimp and the cart was modified such that it contains one or more supported products then the cart must be added to mailchimp with a POST request.

To do this first change the query to get modified quotes in method __getModifiedQuotes()_ of **app/code/community/Ebizmarts/MailChimp/Model/Api/Carts.php** from:
```
        // be sure that the quotes are already in mailchimp and not deleted
        $modifiedCarts->getSelect()->where(
            "m4m.mailchimp_sync_deleted = 0
        AND m4m.mailchimp_sync_delta < updated_at"
        );
```
to
```

        $modifiedCarts->getSelect()->where(
            "m4m.mailchimp_sync_delta < updated_at"

        );
```

and then change the following in that same method from:
```
            $allCarts = $this->addProductNotSentData($mailchimpStoreId, $magentoStoreId, $cart, $allCarts);

            $cartJson = $this->_makeCart($cart, $mailchimpStoreId, $magentoStoreId, true);
            if ($cartJson != "") {
                $this->getHelper()->modifyCounterSentPerBatch(Ebizmarts_MailChimp_Helper_Data::QUO_MOD);

                $counter = $this->getCounter();
                $allCarts[$counter]['method'] = 'PATCH';
                $allCarts[$counter]['path'] = '/ecommerce/stores/' . $mailchimpStoreId . '/carts/'.$cartId;
                $allCarts[$counter]['operation_id'] = $batchId . '_' . $cartId;
                $allCarts[$counter]['body'] = $cartJson;
                $this->setCounter($this->getCounter()+1);
                $this->_updateSyncData(
                    $cartId,
                    $mailchimpStoreId,
                    null,
                    null,
                    0,
                    null,
                    null,
                    $this->getToken()
                );

                /*
                $helper->logDebug(
                    "Creating e-commerce batch for store $magentoStoreId: "
                    . "Adding updated cart ID $cartId for customer email "
                    . $cart->getCustomerEmail() . " currency "
                    . $cart->getQuoteCurrencyCode() . " order_total "
                    . sprintf("%.4f", $cart->getGrandTotal())
                );
                */
            } else {
                $this->_updateSyncData($cartId, $mailchimpStoreId);
            }

            $this->setToken(null);
        }

        return $allCarts;
```
to:
```
            $allCarts = $this->addProductNotSentData($mailchimpStoreId, $magentoStoreId, $cart, $allCarts);

            /*
             * If item was deleted for an emptied cart or cart containing
             * only unsupported products then re-create cart if new cart
             * contains at least one supported product.
             */
            $action = ($cart->getMailchimpSyncDeleted() ? 'POST' : 'PATCH');
            $path = '/ecommerce/stores/' . $mailchimpStoreId . '/carts';
            if ($action == 'PATCH') {
                $path .= '/' . $cartId;
            }

            $cartJson = $this->_makeCart($cart, $mailchimpStoreId, $magentoStoreId, true);
            if ($cartJson != "") {
                $this->getHelper()->modifyCounterSentPerBatch(Ebizmarts_MailChimp_Helper_Data::QUO_MOD);

                $counter = $this->getCounter();
                $allCarts[$counter]['method'] = $action;
                $allCarts[$counter]['path'] = $path;
                $allCarts[$counter]['operation_id'] = $batchId . '_' . $cartId;
                $allCarts[$counter]['body'] = $cartJson;
                $this->setCounter($this->getCounter() + 1);
                $this->_updateSyncData(
                    $cartId,
                    $mailchimpStoreId,
                    null,
                    null,
                    0,
                    null,
                    0,
                    $this->getToken()
                );
                /*
                $actionDesc = ($action == 'POST' ? 'Adding' : 'Modifying');
                $helper->logDebug(
                    "Creating e-commerce batch for store $magentoStoreId: "
                    . "$actionDesc cart ID $cartId for customer email "
                    . $cart->getCustomerEmail() . " currency "
                    . $cart->getQuoteCurrencyCode() . " order_total "
                    . sprintf("%.4f", $cart->getGrandTotal())
                );
                */
            } else {
                /*
                 * Item is empty or has no supported products
                 */
                $reason = $cart->hasItems() ? 'unsupported products only' : 'empty';
                if (!$cart->getMailchimpSyncDeleted()) {
                    /*
                     * Cart previously existed so we need to delete it.
                     */
                    $this->getHelper()->modifyCounterSentPerBatch(Ebizmarts_MailChimp_Helper_Data::QUO_DEL);

                    $counter = $this->getCounter();
                    $allCarts[$counter]['method'] = 'DELETE';
                    $allCarts[$counter]['path'] = '/ecommerce/stores/' . $mailchimpStoreId . '/carts/' . $cartId;
                    $allCarts[$counter]['operation_id'] = $batchId . '_' . $cartId;
                    $allCarts[$counter]['body'] = '';
                    $this->_updateSyncData(
                        $cartId,
                        $mailchimpStoreId,
                        null,
                        null,
                        0,
                        null,
                        1
                    );
                    $this->setCounter($this->getCounter() + 1);
                    /*
                    $helper->logDebug(
                        "Creating e-commerce batch for store $magentoStoreId: "
                        . "Removing cart ID $cartId for "
                        . "customer email " . $cart->getCustomerEmail()
                        . ": " . $reason
                    );
                    */
                } else {
                    /*
                     * Cart was previously empty so we just mark the sync
                     * item as invalid or empty so that it isn't deleted
                     * as an unsent item.
                     */
                    $this->_updateSyncData(
                        $cartId,
                        $mailchimpStoreId,
                        null,
                        null,
                        0,
                        null,
                        null,
                        null,
                        true
                    );
                    /*
                    $helper->logDebug(
                        "Creating e-commerce batch for store $magentoStoreId: "
                        . "Ignoring update for cart ID $cartId for "
                        . "customer email " . $cart->getCustomerEmail()
                        . ": " . $reason
                    );
                    */
                }
            }

            $this->setToken(null);
        }

        return $allCarts;
```
Finally, add the new _QUO_DEL_ batch counter constant to **app/code/community/Ebizmarts/MailChimp/Helper/Data.php** helper:
```
    const QUO_MOD          = "QuoteModified";
    const QUO_DEL          = "QuoteDeleted";
    const QUO_NEW          = "QuoteNew";
```

**Comment by giorello** [comment]
Hi Troy, 

Thanks you again for your helpful and well explained feedback.
We will add this asap.

Kind Regards, 
Javier Giorello.

**Comment by giorello** [comment]
Solved in #1059.
